File: leet_code/graphs/number_of_islands.js
// 0(n) time , O(1) space

// two loops to iterate over grid 
// dfs on all values == 1
// check edge cases so row and col are within bounds
// change 1 into 0 so we don't revisit

var numIslands = function(grid) {
  let count = 0;

  for (let i = 0; i < grid.length; i++) {
    for (let j = 0; j < grid[0].length; j++) {
      if (grid[i][j] == 1) {
        count += traverse(i, j, grid);
      }
    }
  }

  return count;
};

function traverse(i, j, grid) {
  let stack = [[i, j]];
  let neighbors;

  while (stack.length) {
    [i, j] = stack.pop();

    if (grid[i][j] == 0) continue;
    grid[i][j] = 0;

    neighbors = getNeighbors(i, j, grid);

    neighbors.forEach(neighbor => {
      stack.push(neighbor);
    });
  }
  return 1;
}

function getNeighbors(i, j, matrix) {
  neighbors = [];

  if (i > 0) neighbors.push([i - 1, j]);
  if (j > 0) neighbors.push([i, j - 1]);
  if (i < matrix.length - 1) neighbors.push([i + 1, j]);
  if (j < matrix[0].length - 1) neighbors.push([i, j + 1]);

  return neighbors;
}

File: leet_code/python/graphs/number_of_islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        if not grid or len(grid) < 1:
            return 0

        number_of_islands = 0

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == '1':
                    number_of_islands += 1
                    self.sink(grid, i, j)

        return number_of_islands

    def sink(self, grid, i, j):
        if not((0 <= i < len(grid)) and (0 <= j < len(grid[0])) and grid[i][j] == '1'):
            return

        grid[i][j] = '0'

        for node in self.get_neighbors(grid, i, j):
            self.sink(grid, node[0], node[1])

    def get_neighbors(self, matrix, row, col):
        dr = [1, -1, 0, 0]
        dc = [0, 0, 1, -1]

        neighbors = []

        for i in range(len(dr)):
            r = row + dr[i]
            c = col + dc[i]

            if r < 0 or c < 0:
                continue
            if r >= len(matrix) or c >= len(matrix[0]):
                continue
            if matrix[r][c] == 0:
                continue

            neighbors.append([r, c])

        return neighbors
